contrib/mesos/pkg/minion/tasks/task.go

```diff
@@ -275,27 +283,42 @@ func taskRunning(t *Task) taskStateFn {
 	}()
 
 	// wait for the process to complete, or else for a "quit" signal on the task (at which point we'll attempt to kill manually)
-	select {
-	case <-t.shouldQuit:
-		// check for tie
+	gracePeriod := defaultKillGracePeriod
+	forceKill := false
+	wr := &Completion{err: fmt.Errorf("failed to kill process: %q", t.bin)}
+	defer func() { trySend(wr) }() // don't capture wr early
+waitLoop:
+	for i := 0; i < 2; i++ {
 		select {
-		case wr := <-waitCh:
-			// we got a signal to quit, but we're already finished; attempt best effort delvery
-			trySend(wr)
-		default:
-			// Wait() has not exited yet, kill the process
-			log.Infof("killing %s : %s", t.name, t.bin)
-			pid, err := t.cmd.Kill()
+		case <-t.shouldQuit:
+			// check for tie
+			select {
+			case wr = <-waitCh:
+				break waitLoop
+			default:
+			}
+
+			log.Infof("killing %s (force=%t) : %s", t.name, forceKill, t.bin)
+			pid, err := t.cmd.Kill(forceKill)
 			if err != nil {
-				trySend(&Completion{err: fmt.Errorf("failed to kill process: %q pid %d: %v", t.bin, pid, err)})
+				log.Warning("failed to kill process: %q pid %d: %v", t.bin, pid, err)
+				break waitLoop
+			}
+
+			// Wait for the kill to be processed, and child proc resources cleaned up; try to avoid zombies!
+			select {
+			case wr = <-waitCh:
+				break waitLoop
+			case <-time.After(gracePeriod):
+				gracePeriod /= 2
+				forceKill = true
+				continue waitLoop
 			}
-			// else, Wait() should complete and send a completion event
+		case wr = <-waitCh:
+			break waitLoop
 		}
-	case wr := <-waitCh:
-		// task has completed before we were told to quit, pass along completion and error information
-		trySend(wr)
 	}
-	return taskShouldRestart
+	return
```

Comment:
My understanding was that go requires you to do this when you have named the output args.